refactor: restructure/centralize handling of coordinator key selection algorithm

cylewitruk opened the pull request:
## Description

Something that's been nagging at me for a while, so I went ahead and restructured/centralized the "is coordinator pubkey for" logic a little bit.

## Changes

- Moves the core logic into the `keys` module, which feels more natural to me as it's a cross-cutting "key-related" concern (and helps to reduce the size of `transaction_coordinator.rs` slightly).
- Adds tests for the core logic.
- The updated implementation is slightly more flexible by taking generic parameters (`Borrow<bitcoin::BlockHash>` due to its blanket implementation, allowing both `bitcoin::BlockHash` and our `BitcoinBlockHash` and `BitcoinBlockRef` types to be used interchangeably, which is nice in various calling scenarios, and `Into<PublicKey>` so that we can pass most of our pubkey types depending on what you have at the callsite).
- Centralized the "select coordinator" logic for the `SignerInfo` test helper.
- Minor tweaks around the codebase due to the above.

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

cylewitruk committed:
moved helper code here

cylewitruk committed:
apply the actual refactors

djordon commented on signer/src/testing/transaction_signer.rs:
```diff
@@ -10,6 +10,7 @@ use crate::context::SignerEvent;
 use crate::context::SignerSignal;
 use crate::context::TxSignerEvent;
 use crate::error::Error;
+use crate::keys::CoordinatorPublicKey;
```
Turns out that 90% of my comments so far are like this 🤷🏽‍♂️

cylewitruk replied:
Ah, yeah, I've addressed a number of these elsewhere, we're missing quite a few of them.

cylewitruk replied:
Attacked a number of them in [1f993ca](https://github.com/stacks-sbtc/sbtc/pull/1727/commits/1f993ca191d7a28803ac9b0af3db6805905b91ad)

djordon replied:
Awesome, thanks.

cylewitruk commented on signer/src/storage/memory/read.rs:
```diff
@@ -592,7 +592,7 @@ impl DbRead for SharedStore {
         let bitcoin_blocks = std::iter::successors(Some(chain_tip), |block_hash| {
             db.bitcoin_blocks
-                .get(block_hash)
+                .get(*block_hash)
```
I played around with this for a while and it seems to be some sort of compiler trait resolution confusion (hashmaps etc. depend on `Borrow<T>` as well). According to everything that I know and have read it _should_ still work. But it was only a handful of places where the deref from `&&` was needed, so I didn't put too much more effort on it. We could implement an explicit `Borrow<T> for &&` if we really want to avoid the deref, but that feels overkill.

cylewitruk committed:
Merge branch 'main' into feat/consolidate-coordinator-key-handling

cylewitruk committed:
fix up some use statements

cylewitruk commented on signer/src/keys.rs:
```diff
@@ -28,14 +28,18 @@
 use secp256k1::SECP256K1;
 use serde::Deserialize;
 use serde::Serialize;
+use sha2::Digest;
```
[1f993ca](https://github.com/stacks-sbtc/sbtc/pull/1727/commits/1f993ca191d7a28803ac9b0af3db6805905b91ad)

cylewitruk commented on signer/src/testing/wsts.rs:
```diff
@@ -39,6 +41,48 @@ pub struct SignerInfo {
+    for<'a> &'a I: IntoIterator<Item = &'a SignerInfo>,
+{
+    #[track_caller]
+    fn select_coordinator_private_key<B>(&self, bitcoin_chain_tip: B) -> PrivateKey
```
Moved it to the trait fn definition in [1f993ca](https://github.com/stacks-sbtc/sbtc/pull/1727/commits/1f993ca191d7a28803ac9b0af3db6805905b91ad) btw.

cylewitruk committed:
Merge branch 'main' into feat/consolidate-coordinator-key-handling

cylewitruk committed:
Merge branch 'feat/test-helpers-1' into feat/consolidate-coordinator-key-handling

cylewitruk committed:
post-merge fixes

cylewitruk committed:
Merge branch 'main' into feat/consolidate-coordinator-key-handling

cylewitruk committed:
Merge branch 'feat/test-helpers-1' into feat/consolidate-coordinator-key-handling

cylewitruk committed:
Merge branch 'feat/test-helpers-1' into feat/consolidate-coordinator-key-handling

matteojug commented on signer/src/testing/wsts.rs:
```diff
@@ -39,6 +41,48 @@ pub struct SignerInfo {
+/// Trait for selecting the coordinator's private key from a list of items which
+/// contain keypairs.
+pub trait SelectCoordinatorPrivateKey {
+    /// Select the coordinator's private key based on the given list of `SignerInfo`.
```
nit: mentioning `SignerInfo` here seems unexpected

cylewitruk replied:
yeah, missed to update it, [4c577f5](https://github.com/stacks-sbtc/sbtc/pull/1727/commits/4c577f589c91ae212f328ce24cd99f2b7ae28dd3)

matteojug reviewed (approved):
LGTM, just some nits

cylewitruk committed:
feat: additional test helpers & utilities (part 1) (#1726)

## Description

A collection of some smaller test helper/ergonomics improvements. Not
everything is necessarily used here but will be more evident in the next
parts.

## Changes

- Removed some outdated `todo`'s from `default.toml`.
- New, versatile `TestUtilityError` type and error variant to make it
easy to return arbitrary errors from test utilities without needing to
create new error variants.
- New utility/helper traits, types and methods, including
`SleepAsyncExt`, `ResultIterExt`, `FutureExt`, `FuturesIterExt` and
`Sleep`. These are very minor, but provide small ergonomic/verbosity
wins in test code. Refactored
`tests/integration/transaction_coordinator.rs` to show their usages.
- Additional `wait_for_latest_dkg_to_become_verified()` and
`wait_for_key_rotation_event` test-wait helpers.
- Introduce `#[track_caller]` to more test utility functions and having
said methods return better panic messages. This attribute places the
actual panic at the _call site_ of the function instead of within the
utility function itself.
- Changed a number of `.unwrap()`s to panic with better messages.
- Changed a number of `.generate_blocks(1).pop().unwrap()` and similar
usages to `.generate_block()`.
- Other small tweaks/adjustments/etc.

## Checklist:

- [x] I have performed a self-review of my code
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream
modules

djordon commented on signer/src/keys.rs:
```diff
@@ -476,11 +480,82 @@ impl SignerScriptPubKey for secp256k1::XOnlyPublicKey {
 
+/// Utility methods for determining the coordinator public key based on the
+/// underlying set of signers and the bitcoin chain tip.
+pub trait CoordinatorPublicKey {
+    /// Find the coordinator public key
+    fn determine_for<B, K>(
+        bitcoin_chain_tip: B,
+        signer_public_keys: &BTreeSet<K>,
+    ) -> Option<PublicKey>
```
Why tie this to the trait instead instead of having this as a free function? Also, I can see why making it generic over `B` is useful, but why make it generic over `K`?

cylewitruk replied:
> Why tie this to the trait instead instead of having this as a free function?

Just a preference of taste, I suppose. `CoordinatorPublicKey::determine_for(tip, keys)` gives nice and encapsulated context (even though `BTreeSet::<PublicKey>::determine_for(tip, keys)` is a bit ambiguous, it's not very often one calls associated methods on the collection type like that, and intellisense anyway highlights that it's from the `CoordinatorPublicKey` trait). 🤷 

> Also, I can see why making it generic over B is useful, but why make it generic over K?

No strong argument, largely just ergonomics as we sometimes use `secp256k1` etc. types directly in tests, so you can call `CoordinatorPublicKey::determine_for()` syntax if you've got a set of those and want to find the coordinator key. This will result in something like 5 concrete impls, but optimized they should be very small and/or optimized away if unused, and only the final`nth` key is copied and mapped, so no real measurable performance impact.

djordon commented on signer/src/keys.rs:
```diff
@@ -689,4 +764,124 @@ mod tests {
+        bytes[0] = seed; // Simple differentiation
+        // For more varied hashes, could use a simple counter or fill more bytes.
+        // For example, to make it more "hash-like" for testing distribution:
+        let mut full_seed = [0u8; 32];
+        full_seed[0] = seed;
+        full_seed[1] = seed.wrapping_add(1);
+        full_seed[2] = seed.wrapping_add(2);
+        // ... and so on, or use a seeded RNG to fill `bytes`.
+        // For this test, simple differentiation is likely sufficient.
```
Why not sha256 the seed? Just curious, what's here is fine too.

cylewitruk replied:
`full_seed` was dead code so I just removed it, the above was sufficient for the simple `u8` differentiator (and makes it easy to see which seed caused any potential issues directly in the block hash). [c8db5e9](https://github.com/stacks-sbtc/sbtc/pull/1727/commits/c8db5e938b17effb2ab1211414b257afd5218064)

MCJOHN974 committed:
[chore]: fix some markdown lint issues (#1758)

## Description

Closes: no issue

I fixed some formatting in our markdown files according to markdown
standard described in https://github.com/DavidAnson/markdownlint

## Notes

- Almost all changes I did was machine-applicable.
- This markdown standard is not an official standard, but afaik it is
the most popular one.
- I didn't fix some of the issues, mostly in issue templates since I'm
not sure if github is ok with them.

djordon committed:
fix: specify the event name in the wait-list of wait-other-jobs (#1762)

## Description

This may fix that issue that we've been seeing in CI. Doesn't address
the warnings though.

The underlying action can be found here
https://github.com/kachick/wait-other-jobs/tree/215c93f93f91e8f9a5ccb7fc0cc78220c9fd8a99?tab=readme-ov-file#usage,
while our use of it can be found here
https://github.com/stacks-sbtc/actions/blob/181f8c67da2707c66b5e31f24e7418c47adefdd1/wait-other-jobs/action.yml

## Changes

1. Specify the `eventName` in the wait list, in the same way that the
usage docs demonstrate. But I'm not 100% sure that this fixes it, also
not sure whether this introduces other issues.

## Testing Information

## Checklist

- [x] I have performed a self-review of my code

djordon committed:
chore: use transactions in database migrations (#1761)

## Description

This is basically https://github.com/stacks-sbtc/sbtc/pull/1670, just
with signed commits.

## Changes

* Apply the migrations within a transaction.

## Testing Information

If tests pass then we should be good.

## Checklist:

- [x] I have performed a self-review of my code

MCJOHN974 committed:
[feat]: add RBF status to Emily (#1686)

## Description

Closes: #1685

## Changes
- Now Emily have 1 more event status -- RBF.
- For RBF events Emily have `replaced_by_tx` field. It stores txid of
transaction which replaced this transaction during an RBF.

*Note:* this PR solves only Emily side of things, for sidecar changes
I'd add a new one.

## Testing Information
- added baseline happy path test
- updated testcases for permissions and duplicates tests with RBF
testcase.
- added test checking that only events with RBF status can have
`replaced_by_tx` field.

## Checklist:

- [x] I have performed a self-review of my code

---------

Co-authored-by: Francesco <2530388+Jiloc@users.noreply.github.com>
Co-authored-by: Matteo Almanza <matteojug@gmail.com>
Co-authored-by: Daniel Jordon <djordon@users.noreply.github.com>

djordon committed:
fix: support brand new bootstrap signer set (#1752)

## Description

Closes https://github.com/stacks-sbtc/sbtc/issues/1751

## Changes

1. Load the aggregate key from the last DKG shares if we did not run DKG
and do not have any signer set info in the registry.
2. Signers proceed with verification using the latest verified DKG
shares if they do not have info in the registry. They fall back to the
latest shares if this is a contract deployment and the contracts have
not been deployed.

## Testing Information

Tested this on devenv with a fresh build and was able to sweep a deposit
without any issues.

Also, I redid one of our previously ignored tests and completely removed
two of the variants that were ignored. I couldn't get them to work and
got tired of attempting to fix them. Perhaps it would be better to keep
them around but ignored. I'm okay either way.

## Checklist:

- [x] I have performed a self-review of my code

MCJOHN974 committed:
chore: Split deposit and withdrawal status on Emily (#1745)

## Description

Closes: #1693

## Changes

Now Emily have two structs `DepositStatus` and `WithdrawalStatus`
instead of old `Status` (same for `StatusEntry`). As for now they differ
because deposits can have `RBF` status while withdrawals can't, and this
split protect us from invalid state (withdrawal with `RBF` status) on
compiler level.

## Testing Information

Refactoring. No new testing needed

## Checklist:

- [x] I have performed a self-review of my code

---------

Co-authored-by: Francesco <2530388+Jiloc@users.noreply.github.com>
Co-authored-by: Matteo Almanza <matteojug@gmail.com>

MCJOHN974 committed:
chore: remove reprocessing status from Emily (#1744)

## Description

Closes: #1678

## Changes

Removed reprocessing status from Emily and Emily tests, we don't use it
and don't plan to use it

## Testing Information

No new tests needed

## Checklist:

- [x] I have performed a self-review of my code

---------

Co-authored-by: Francesco <2530388+Jiloc@users.noreply.github.com>
Co-authored-by: Matteo Almanza <matteojug@gmail.com>

cylewitruk committed:
Merge branch 'main' into feat/consolidate-coordinator-key-handling

cylewitruk committed:
remove dead test helper code

cylewitruk committed:
fix comment

djordon-stacks commented:
Closing since we haven't seen much action pushing it through, and we are unlikely to see that anytime soon. Plus, we can always reopen it later.
